conclave-enclave/src/main/kotlin/com/r3/conclave/enclave/internal/NativeEnclaveHostInterface.kt

```diff
@@ -0,0 +1,120 @@
+package com.r3.conclave.enclave.internal
+
+import com.r3.conclave.common.EnclaveStartException
+import com.r3.conclave.common.internal.*
+import com.r3.conclave.mail.MailDecryptionException
+import com.r3.conclave.utilities.internal.getAllBytes
+import java.nio.ByteBuffer
+import kotlin.collections.ArrayDeque
+
+typealias StackFrame = CallInterfaceStackFrame<HostCallType>
+
+/**
+ * This class is the implementation of the [EnclaveHostInterface] for native enclaves.
+ * It has three jobs:
+ *  - Serve as the endpoint for calls to make to the host, see [com.r3.conclave.common.internal.CallInterface]
+ *  - Route calls from the host to the appropriate enclave side call handler, see [com.r3.conclave.common.internal.CallInterface]
+ *  - Handle the low-level details of the messaging protocol (ECalls and OCalls).
+ */
+class NativeEnclaveHostInterface : EnclaveHostInterface() {
+    /** In release mode we want to sanitise exceptions to prevent leakage of information from the enclave */
+    var sanitiseExceptions: Boolean = false
+
+    /**
+     * Each thread has a lazily created stack which contains a frame for the currently active host call.
+     * When a message arrives from the host, this stack is used to associate the return value with the corresponding call.
+     */
+    private val threadLocalStacks = ThreadLocal.withInitial { ArrayDeque<StackFrame>() }
+    private val stack get() = threadLocalStacks.get()
+
+    private fun checkCallType(type: HostCallType) = check(type == stack.last().callType) { "Call type mismatch" }
+
+    /**
+     * Internal method for initiating a host call with specific arguments.
+     * This should not be called directly, but instead by implementations in [EnclaveHostInterface].
+     */
+    override fun executeOutgoingCall(callType: HostCallType, parameterBuffer: ByteBuffer): ByteBuffer? {
+        stack.addLast(StackFrame(callType, null, null))
+
+        Native.jvmOCall(
+                callType.toByte(), CallInterfaceMessageType.CALL.toByte(), parameterBuffer.getAllBytes(avoidCopying = true))
+
+        val stackFrame = stack.removeLast()
```

Comment:
Is `stackFrame` the same as the one that was added at the beginning of the method? Should we have an instance equality (`===`) check for sanity?

Reply:
It is, or at least it should be. Added just to ensure nothing goes awry.